base/linalg/triangular.jl

```diff
@@ -763,69 +763,60 @@ function A_mul_Bc!(A::StridedMatrix, B::UnitLowerTriangular)
 end
 
 #Generic solver using naive substitution
-function naivesub!(A::UpperTriangular, b::AbstractVector, x::AbstractVector=b)
+# manually hoisting x[j] significantly improves performance as of Dec 2015
+# manually eliding bounds checking significantly improves performance as of Dec 2015
+# directly indexing A.data rather than A significantly improves performance as of Dec 2015
+function naivesub!(A::UpperTriangular, b::AbstractVector, x::AbstractVector = b)
     n = size(A, 2)
     if !(n == length(b) == length(x))
-        throw(DimensionMismatch("second dimension of left hand side A, $n, length of output x, $(length(x)), and length of right hand side b, $(length(b)) must be equal"))
+        throw(DimensionMismatch("second dimension of left hand side A, $n, length of output x, $(length(x)), and length of right hand side b, $(length(b)), must be equal"))
     end
-    for j = n:-1:1
-        xj = b[j]
-        for k = j+1:1:n
-            xj -= A[j,k] * x[k]
-        end
-        Ajj = A[j,j]
-        if Ajj == zero(Ajj)
-            throw(SingularException(j))
-        else
-            x[j] = Ajj\xj
+    @inbounds for j in n:-1:1
+        A.data[j,j] == zero(A.data[j,j]) && throw(SingularException(j))
```

Comment:
I thought so as well. So I benchmarked. No significant performance impact. I suppose the compiler is intelligent enough to perform that optimization automagically?

For benefit of future readers, alongside the other optimization notes I added sentences addressing this optimization and manually hoisting A.data out of the loops.

Thanks!